File: app/src/main/java/com/java_programming/pattern/Singleton.java
package com.java_programming.pattern;

/**
 * Created by 1000292 on 15-10-2022
 * BFDL
 * PUNE.
 */

/**
 * https://medium.com/swlh/a-complete-guide-on-singleton-design-pattern-in-java-846c2f8df5fc
 */
class Singleton {

    private static volatile Singleton instance;
    private Singleton() {}
    
    public static Singleton getInstance(){
        if(instance != null)
        {
            return instance;
        }

        synchronized (Singleton.class) {
            if(instance == null)
            {
                instance = new Singleton();
            }
        }
        return instance;

    }

    public static void main(String[] args) {
          System.out.println( Singleton.getInstance());
    }
}

File: app/src/main/java/com/designpattern/creational/Singleton.kt
package com.designpattern.creational

/**
 * Created by 1000292 on 02-11-2022
 * BFDL
 * PUNE.
 */
object Singleton

/*
class Singleton private constructor() {

    companion object {

        @Volatile
         private var instance: Singleton? = null // Volatile modifier is necessary

        fun getInstance() =
            instance ?: synchronized(this) { // synchronized to avoid concurrency problem
                instance ?: Singleton().also { instance = it }
            }
    }
}*/

// href=https://medium.com/@ZahraHeydari/singleton-pattern-in-kotlin-b09380c53b14

/*class Singleton private constructor() {
    companion object {
        val INSTANCE: Singleton by lazy(LazyThreadSafetyMode.SYNCHRONIZED) { Singleton() }
    }
}*/


/*fun main() {
    println(Database())
    println(Database())
}

class Database private constructor(){
    companion object{
        private var instance : Database ? = null

        operator fun invoke ()  = synchronized(this){
            if(instance==null)
                instance = Database()
            instance
        }
    }
}*/


/*class UsersDatabase : RoomDatabase() {

    companion object {

        @Volatile private var INSTANCE: UsersDatabase? = null

        fun getInstance(context: Context): UsersDatabase =
            INSTANCE ?: synchronized(this) {
                INSTANCE ?: buildDatabase(context).also { INSTANCE = it }
            }

        private fun buildDatabase(context: Context) =
            Room.databaseBuilder(context.applicationContext,
                UsersDatabase::class.java, "Sample.db")
                .build()
    }
}*/
